src/lib/optimizer/strategy/predicate_merge_rule.cpp

```diff
@@ -0,0 +1,105 @@
+#include "predicate_merge_rule.hpp"
+
+#include "expression/expression_functional.hpp"
+#include "expression/logical_expression.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/union_node.hpp"
+
+using namespace opossum::expression_functional;  // NOLINT
+
+namespace opossum {
+
+PredicateMergeRule::PredicateMergeRule(const size_t optimization_threshold)
+    : _optimization_threshold(optimization_threshold) {}
+
+void PredicateMergeRule::apply_to(const std::shared_ptr<AbstractLQPNode>& root) const {
+  Assert(root->type == LQPNodeType::Root, "PredicateMergeRule needs root to hold onto");
+
+  visit_lqp(root, [&](const auto& node) {
+    size_t lqp_complexity = 0;
+    visit_lqp(node, [&](const auto& sub_node) {
```

Comment:
Two nested traversal of DAGs are hard to understand / verify. Can you think of a way to do this in a single pass? Also, "complexity" sounds like something more sophisticated than what we are doing right now.